### PR TITLE
Fix undefined showing up in output of hexFormat

### LINKED ISSUE #5: Id contains undefined
I'm sometimes getting ids that look like this:

`59a044090004d2undefined13c3c000001`

I think the issues is that sometimes hexFormat is called with a width that is smaller than the input string.

### PR DESCRIPTION
**Pull request by cwohlman**
Fixes #5 

Note, on my machine there are still 2 failing tests:

```
AssertionError: false is 'truthy'
    at Object.id should include pid (/Users/joshuaohlman/dev/node-mongoid-js/test/test-mongoid.js:155:11)
    at /Users/joshuaohlman/dev/node-mongoid-js/node_modules/qnit/lib/qnit.js:397:32
    at /Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:109:13
    at _invokeCallback (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:49:20)
    at applyFunc (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:103:50)
    at _loop (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:62:13)
    at repeatUntil (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:72:5)
    at Object.applyVisitor (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:101:5)
    at runit (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/qnit/lib/qnit.js:369:15)
    at runTestFunction (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/qnit/lib/qnit.js:352:5)

 X MongoId_class - testShouldParseId (1.115ms)

AssertionError: 5700 == 91208
    at Object.testShouldParseId (/Users/joshuaohlman/dev/node-mongoid-js/test/test-mongoid.js:194:14)
    at /Users/joshuaohlman/dev/node-mongoid-js/node_modules/qnit/lib/qnit.js:397:32
    at /Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:109:13
    at _invokeCallback (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:49:20)
    at applyFunc (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:103:50)
    at _loop (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:62:13)
    at repeatUntil (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:72:5)
    at Object.applyVisitor (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/aflow/lib/qflow.js:101:5)
    at runit (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/qnit/lib/qnit.js:369:15)
    at runTestFunction (/Users/joshuaohlman/dev/node-mongoid-js/node_modules/qnit/lib/qnit.js:352:5)

```

### CODE REVIEW

File: mongoid.js [hunk 1]
```diff
@@ -109,7 +109,12 @@ MongoId.prototype.mongoid = MongoId.prototype.fetch;
 var _zeroPadding = ["", "0", "00", "000", "0000", "00000", "000000", "0000000"];
 function hexFormat(n, width) {
     var s = n.toString(16);
-    return _zeroPadding[width - s.length] + s;
+    var paddingNeeded = width - s.length;
+    if (paddingNeeded > 0) {
```
Comment (andrasq) [inline mongoid.js:113]: this could be simpler written as `if (s.length <= width)`

File: mongoid.js [hunk 2]
```diff
@@ -109,7 +109,12 @@ MongoId.prototype.mongoid = MongoId.prototype.fetch;
...
+        return _zeroPadding[width - s.length] + s;
+    } else {
+        return s.substr(0, width);
```
Comment (andrasq) [inline mongoid.js:116]: for most id fields keeping the least significant bits would preserve uniqueness better (ie, for all the incrementing fields like process id), so maybe `return s.substr(-width)`.  In this case it is the process id that was too long.

Note that with a clipped process id it's not possible to guarantee uniqueness (well, one could formulate a "machine id" that preserved the process uniqueness I suppose)

**Review by andrasq — CHANGES_REQUESTED** [review]
the tests are failing because is uses the actual process id which didn't fit into the generated mongoid.  When the id is clipped with `substr` the original id will not be found in the mongoid, nor is it recoverable with parseId.

Should also have a test to check that the change is working (ie, that the generated mongoid is valid when the process id is longer than 16 bits)

**Comment by cwohlman** [comment]
Thanks for the feedback. I'll update the PR when I get a chance.

Question: do you think there's any other way to solve the problem than clipping the process id?

**Comment by andrasq** [comment]
the problem is that a mongoid uses 16 bits to store the process id, but it seems some systems use process ids longer than 16 bits.  Clipping is necessary for a correctly formed id.

The extra bits can be discarded (clipped), they can be disallowed (throw an error if the process id is too big), or can be stuffed into one of the other fields.  Stuffing would make a non-standard mongoid, but would parse as valid and would preserve uniqueness.

Stuffing the extra bits into the machine id wouldn't even require changing the code, just a properly constructed system id to the constructor, eg:

```
// the 4 high bits of the process id followed by 20 random bits
// 16-bit process ids will produce machine ids with a leading 0 hex digit
const machineId = ((process.pid & 0xF0000) << 4) + Math.floor(Math.random() * (1 << 20));
const mongoid = new Mongoid(machineId);
```

**Comment by andrasq** [comment]
the split process id and hacked system id handling could be moved into mongoid-js I guess, I'll have to think about that one.  parse() would have to be aware of it too, but none of the changes should affect the critical path.

File: mongoid.js (hunk 1 above)
Comment (andrasq) [reply mongoid.js:113]: or even simpler if the incoming process id is masked first, ie
`this.processIdStr = hexFormat(machineId, 6) + hexFormat(processId & 0xFFFF, 4);`
(the other formatted fields, timestamp, sequence and machineId, are already length checked)

**Comment by andrasq** [comment]
This won't work in all cases.  The linux `kernel.pid_max` can be configured up to 31 bits, whereas the trick above only saves 4 more bits.

**Comment by cwohlman** [comment]
Hm... I guess I'll just make the changes you originally suggested for now. It sounds like trying to support larger process ids will mean more complexity without fully solving the problem.

**Comment by andrasq** [comment]
I merged in PR #7 that should fix the broken ids that included "undefined" in the id string and fixes the unit tests that sometimes failed on systems with 17 bit and longer process ids.

A more "clever" way of supporting longer process ids would be more complex, would not be portable (ie only mongoid-js would know how to recover the full pid from an id string), and would only add a few additional bits of process id precision.  The linux `kernel.pid_max` can be configured to max 2 ^ 22, ie 22 bits.

**Comment by cwohlman** [comment]
Closing since #7 fixes this issue.

**Pull request closed without merging**